$\large{\textbf{Solution to Question 1:}}$

In [2]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 5.1 MB/s 
     |████████████████████████████████| 49 kB 5.4 MB/s 


In [3]:
from pyomo.environ import *

In [4]:
import numpy as np

In [5]:
#Create a model
model1 = ConcreteModel()

In [6]:
#Declaring the no of variables
N = 14

In [7]:
#Creating array of coefficients of objective function
obj_coeff_c = np.array([ -4.2, 37.75, 3.5, 10.05, -8.25, 40.85, 0.05, -11.12, 0.65, 20.725, -102.65, 0.6, -9.45, -25.625])

In [8]:
#Declaring the arrays for constraints
constr2_coef_w = np.array([2.8, -0.6725, 3.6, 0, -2.02, 1.01, -4.305, 0, 1, -2.09, 0, -2.124, -3.402, 0])

#declare the RHS of the constraints as variables
constr1_rhs_b1 = 343.48
constr2_rhs_b2 = 66.2

In [9]:
#The following numpy arrays will be useful for setting the lower bound and upper bounds for the variables
lower_bound = np.array([2, -3, -np.inf, 1, 7, 3, -9, 0, 1, -25.245, -np.inf, -6, -30.5, -10.8])
upper_bound = np.array([np.inf, 6, 2.5, 3.4, np.inf, 6, 4.1, np.inf, 30, 11, 0, 0, 95, 13.4])

In [10]:
# set of column indices:
col_indices = np.arange(N)

In [11]:
#declare the decision variables in the model
model1.x = Var(col_indices)

In [12]:
# create a ConstraintList to hold multiple constraints
model1.constraints = ConstraintList()

In [13]:
#Adding model constraints to the list
model1.constraints.add(sum(model1.x[j] for j in col_indices) <= constr1_rhs_b1)
model1.constraints.add(sum(constr2_coef_w[j]*model1.x[j] for j in col_indices) == constr2_rhs_b2)

In [14]:
#add the upper bound and lower bounds for the variables
for j in col_indices:
  model1.x[j].setlb(lower_bound[j])
  model1.x[j].setub(upper_bound[j])

In [15]:
model1.objective = Objective(expr = sum(obj_coeff_c[j]*model1.x[j] for j in col_indices), sense=minimize)

In [16]:
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   14 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}

1 Var Declarations
    x : Size=14, Index=x_index
        Key : Lower   : Value : Upper : Fixed : Stale : Domain
          0 :     2.0 :  None :  None : False :  True :  Reals
          1 :    -3.0 :  None :   6.0 : False :  True :  Reals
          2 :    None :  None :   2.5 : False :  True :  Reals
          3 :     1.0 :  None :   3.4 : False :  True :  Reals
          4 :     7.0 :  None :  None : False :  True :  Reals
          5 :     3.0 :  None :   6.0 : False :  True :  Reals
          6 :    -9.0 :  None :   4.1 : False :  True :  Reals
          7 :     0.0 :  None :  None : False :  True :  Reals
          8 :     1.0 :  No

In [17]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [18]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -14246.25704375
  Upper bound: -14246.25704375
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 15
  Number of nonzeros: 25
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.015396356582641602
# ----------------------------------------------------------
#   Solution Information
# --------------

In [19]:
# display solution
print('\nObjective = ', model1.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model1.x[i].value)

print('\nConstraints')
model1.constraints.display()


Objective =  -14246.257043750004

Decision Variables
x[ 0 ] =  2039.2619375
x[ 1 ] =  -3.0
x[ 2 ] =  -1623.2369375
x[ 3 ] =  1.0
x[ 4 ] =  7.0
x[ 5 ] =  3.0
x[ 6 ] =  -9.0
x[ 7 ] =  0.0
x[ 8 ] =  1.0
x[ 9 ] =  -25.245
x[ 10 ] =  0.0
x[ 11 ] =  -6.0
x[ 12 ] =  -30.5
x[ 13 ] =  -10.8

Constraints
constraints : Size=2
    Key : Lower : Body               : Upper
      1 :  None : 343.47999999999985 : 343.48
      2 :  66.2 :  66.19999999999925 :   66.2


$\large{\textbf{Answers.}}$

$\textbf{Objective Function Value / Cost = }-14246.257 \\ \textbf{Value of decision variables:} \\ x_1 = 2039.26 \\ x_2 = -3 \\ x_3 = -1623.2369 \\ x_4 = 1 \\ x_5 = 7 \\ x_6 = 3 \\ x_7 = -9 \\ x_8 = 0 \\ x_9 = 1 \\ x_{10} = -25.425 \\ x_{11} = 0 \\ x_{12} = -6 \\ x_{13} = -30.5 \\ x_{14} = -10.8$